packages/app/src/platformLoggerService.ts

```diff
@@ -0,0 +1,49 @@
+/**
+ * @license
+ * Copyright 2019 Google Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { ComponentContainer, ComponentType } from '@firebase/component';
+import { VersionService } from './version-service';
+import { PLATFORM_LOG_STRING } from './constants';
+
+export class PlatformLoggerService {
+  constructor(private readonly container: ComponentContainer) {}
+  // In initial implementation, this will be called by installations on
+  // auth token refresh, and installations will send this string.
+  getPlatformInfoString(): string {
+    const providers = this.container.getProviders();
+    // Loop through providers and get library/version pairs from any that are
+    // version components.
+    return providers
+      .map(provider => {
+        const service = provider.getImmediate() as VersionService;
```

Comment:
`provider.getImmediate()` will throw in case of placeholder providers. You can call it with `optional` flag to make `getImmediate()` return `null` instead of throwing. Then check if service is `null`. 

```
const service = provider.getImmediate({optional: true}) as VersionService;
```

We can also remove the type cast by using a custom type guard, but it needs a hack:
```
provider => {

        if (isVersionServiceProvider(provider)) {
          const service = provider.getImmediate();
          // TODO: We can use this check to whitelist strings when/if we set up
          // a good whitelist system.
          const platformString =
            PLATFORM_LOG_STRING[
            service.library as keyof typeof PLATFORM_LOG_STRING
            ] ?? service.library;
          return `${platformString}/${service.version}`;
        } else {
          return null;
        }
      }

/**
 * 
 * @param provider check if this provider provides a VersionService
 * 
 * NOTE: Using Provider<'app-version'> is a hack to indicate that the provider provides VersionService
 * The provider is not necessarily a 'app-version' provider.
 */
function isVersionServiceProvider(provider: Provider<Name>): provider is Provider<'app-version'> {
  const component = provider.getComponent();
  return component?.type === ComponentType.VERSION;
}
```

Reply:
After switching to this we don't need `Provider.getComponent()` anymore. Should I get rid of it or do we anticipate it being useful at some point?

Reply:
Done.